packages/frontend/core/src/modules/docs-search/entities/docs-indexer.ts

```diff
@@ -83,380 +85,214 @@ export class DocsIndexer extends Entity {
-
-    if (!rootDocBuffer) {
-      return;
-    }
-
-    const yRootDoc = new YDoc();
-    applyUpdate(yRootDoc, rootDocBuffer);
-
-    const docStoragePossibleIds = Array.from(yRootDoc.getSubdocs())
-      .map(doc => doc.guid)
-      .filter(id => id.endsWith(docId));
-
-    let docBuffer;
-
-    for (const id of docStoragePossibleIds) {
-      docBuffer = await this.workspaceEngine.doc.storage.loadDocFromLocal(id);
+    let workerOutput;
 
-      if (docBuffer) {
-        break;
-      }
-    }
-
-    if (!docBuffer) {
-      return;
-    }
+    if (docId === this.workspaceId) {
+      const rootDocBuffer =
+        await this.workspaceEngine.doc.storage.loadDocFromLocal(
+          this.workspaceId
+        );
 
-    const ydoc = new YDoc();
-
-    applyUpdate(ydoc, docBuffer);
-
-    let docExists: boolean | null = null;
-
-    (
-      yRootDoc.getMap('meta').get('pages') as YArray<YMap<any>> | undefined
-    )?.forEach(page => {
-      if (page.get('id') === docId) {
-        docExists = !(page.get('trash') ?? false);
+      if (!rootDocBuffer) {
+        return;
       }
-    });
-
-    if (!docExists) {
-      const indexWriter = await this.docIndex.write();
-      indexWriter.delete(docId);
-      await indexWriter.commit();
 
-      const blockIndexWriter = await this.blockIndex.write();
-      const oldBlocks = await blockIndexWriter.search(
-        {
-          type: 'match',
-          field: 'docId',
-          match: docId,
-        },
-        {
-          pagination: {
-            limit: Number.MAX_SAFE_INTEGER,
+      const allIndexedDocs = (
+        await this.docIndex.search(
+          {
+            type: 'all',
           },
-        }
-      );
-      for (const block of oldBlocks.nodes) {
-        blockIndexWriter.delete(block.id);
-      }
-      await blockIndexWriter.commit();
+          {
+            pagination: {
+              limit: Number.MAX_SAFE_INTEGER,
+              skip: 0,
+            },
+          }
+        )
+      ).nodes.map(n => n.id);
+
+      workerOutput = await worker.run({
+        type: 'rootDoc',
+        allIndexedDocs,
+        rootDocBuffer,
+      });
     } else {
-      const blocks = ydoc.getMap<any>('blocks');
+      const rootDocBuffer =
+        await this.workspaceEngine.doc.storage.loadDocFromLocal(
+          this.workspaceId
+        );
 
-      if (blocks.size === 0) {
+      const docBuffer =
+        (await this.workspaceEngine.doc.storage.loadDocFromLocal(
+          storageDocId
+        )) ?? new Uint8Array(0);
+
+      if (!rootDocBuffer) {
         return;
       }
 
-      let docTitle = '';
-
-      const blockDocuments: Document<typeof blockIndexSchema>[] = [];
-
-      for (const block of blocks.values()) {
-        const flavour = block.get('sys:flavour')?.toString();
-        const blockId = block.get('sys:id')?.toString();
+      workerOutput = await worker.run({
+        type: 'doc',
+        docBuffer,
```

Comment:
is worker available to access indexdb directly?

Reply:
The storage implementation is injected by DI, probably SQLite

Reply:
if we put the whole job & indexing in worker, we can lowest the data transfer cost. the doc buffer can reaches MB sizes